File: java/src/main/java/org/npb/agview/api/example/PremPostHandler.java
package org.npb.agview.api.example;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.Collection;
import java.util.HashMap;
import java.util.function.Function;
import java.util.stream.Collectors;

public class PremPostHandler {

    private final HttpClient httpClient;
    private final Arguments connectionInfo;
    private final AccessTokenHandler accessTokenHandler;
    private final PremDbHandler premDbHandler;

    public PremPostHandler(HttpClient httpClient,
                           Arguments connectionInfo,
                           AccessTokenHandler accessTokenHandler,
                           PremDbHandler premDbHandler) {
        this.httpClient = httpClient;
        this.connectionInfo = connectionInfo;
        this.accessTokenHandler = accessTokenHandler;
        this.premDbHandler = premDbHandler;
    }

    public Collection<CreatedPrem> createPrems() {
        try {
            var prems = premDbHandler.getPremsToLoad();
            var premsAddresses = premDbHandler.getPremAddressesToLoad();

            var premAddressByUsdaPin = premsAddresses.stream().collect(Collectors.toMap(PremAddress::getUsdaPin, Function.identity()));

            var postRequestBody = prems.stream().map(prem ->
                    new HashMap<String, Object>() {
                        {
                            put("usdaPin", prem.getUsdaPin());
                            put("premName", prem.getPremName());
                            put("species", prem.getSpecies());
                            put("iceContactPhone", prem.getIceContactPhone());
                            put("iceContactEmail", prem.getIceContactEmail());
                            put("locationType", prem.getLocationType());
                            put("siteCapacityNumberBarns", prem.getSiteCapacityNumberBarns());
                            put("siteCapacityNumberAnimals", prem.getSiteCapacityNumberAnimals());
                            put("numberOfAnimalsOnSite", prem.getNumberOfAnimalsOnSite());
                            put("streetAddress", premAddressByUsdaPin.get(prem.getUsdaPin()).getStreetAddress());
                            put("city", premAddressByUsdaPin.get(prem.getUsdaPin()).getCity());
                            put("state", premAddressByUsdaPin.get(prem.getUsdaPin()).getState());
                            put("zip", premAddressByUsdaPin.get(prem.getUsdaPin()).getZip());
                        }
                    }
            )
                    .collect(Collectors.toList());

            var objectMapper = new ObjectMapper();
            var bodyJsonStr = objectMapper
                    .writerWithDefaultPrettyPrinter()
                    .writeValueAsString(postRequestBody);
            var accessToken = accessTokenHandler.getNewAccessToken();
            var request = HttpRequest.newBuilder()
                    .uri(new URI(connectionInfo.getBaseUrl() + "/api/v1/prems/"))
                    .headers("Content-Type", "application/json",
                            "Authorization", "Bearer "+accessToken.getAccess())
                    .POST(HttpRequest.BodyPublishers.ofString(bodyJsonStr))
                    .build();
            var future = httpClient.sendAsync(request, HttpResponse.BodyHandlers.ofString())
                            .thenApply(HttpResponse::body);
            var responseObjectMapper =
                    new ObjectMapper().configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);

            return responseObjectMapper.readValue(future.get(), new TypeReference<>(){});
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

}

File: js/src/PremPostHandler.ts
import axios, { AxiosResponse } from 'axios';
import { getData } from "./DbHandler"
import { getNewAccessToken } from "./AccessTokenHandler"
import Constants from './Constants'

export async function createPrems(premFilePath: string, premAddressFilePath: string) {
    let prems = (await getData(premFilePath)).body
    let premAddresses = (await getData(premAddressFilePath)).body

    let premAddressByUsdaPin: Map<string, Map<string, any>> = new Map()
    premAddresses.forEach(premAddress => {
        premAddressByUsdaPin.set(premAddress.get('usda_pin'), premAddress)
    })

    let requestBody: Array<any> = createRequestBody(prems, premAddressByUsdaPin)

    let accessToken = await getNewAccessToken(Constants.NPB_BASE_URL, Constants.NPB_API_KEY, Constants.NPB_API_SECRET)

    return axios.post(`${Constants.NPB_BASE_URL}/api/v1/prems/`, requestBody,
        {
            headers: { Authorization: `Bearer ${accessToken.access}` }
        })
        .then((res: AxiosResponse<any>) => res)
        .catch(error => {
            console.error(error)
        })
}

function createRequestBody(prems: Map<string, any>[], premAddressByUsdaPin: Map<string, Map<string, any>>): Array<any> {
    let requestBody: Array<any> = []
    prems.forEach(prem => {
        let requestBodyEntry: any = {}
        requestBodyEntry['usdaPin'] = getOrNull(prem.get('usda_pin'))
        requestBodyEntry['premName'] = getOrNull(prem.get('prem_name'))
        requestBodyEntry['species'] = getOrNull(prem.get('species'))
        requestBodyEntry['iceContactPhone'] = getOrNull(prem.get('ice_contact_phone'))
        requestBodyEntry['iceContactEmail'] = getOrNull(prem.get('ice_contact_email'))
        requestBodyEntry['locationType'] = getOrNull(prem.get('location_type'))
        requestBodyEntry['siteCapacityNumberBarns'] = getOrNull(prem.get('site_capacity_number_barns'))
        requestBodyEntry['siteCapacityNumberAnimals'] = getOrNull(prem.get('site_capacity_number_animals'))
        requestBodyEntry['numberOfAnimalsOnSite'] = getOrNull(prem.get('number_of_animals_on_site'))
        requestBodyEntry['streetAddress'] = getOrNull(getCorrespondingAddressRecord(premAddressByUsdaPin, prem).get('street_address'))
        requestBodyEntry['city'] = getOrNull(getCorrespondingAddressRecord(premAddressByUsdaPin, prem).get('city'))
        requestBodyEntry['state'] = getOrNull(getCorrespondingAddressRecord(premAddressByUsdaPin, prem).get('state'))
        requestBodyEntry['zip'] = getOrNull(getCorrespondingAddressRecord(premAddressByUsdaPin, prem).get('zip'))

        requestBody.push(requestBodyEntry)
    })

    return requestBody
}

function getCorrespondingAddressRecord(premAddressByUsdaPin: Map<string, Map<string, any>>, prem: Map<string, any>): Map<string, any> {

    return premAddressByUsdaPin.get(prem.get('usda_pin'))!
}

function getOrNull(value: any) {
    return value ? value : null
}
